raftengine/engine.go

```diff
@@ -106,51 +110,71 @@ type stateOp struct {
 }
 
 func (e *Engine) Write(wb *WriteBatch) error {
+	e.writeMu.Lock()
+	defer e.writeMu.Unlock()
 	if int64(wb.size)+e.writer.offset() > e.writer.walSize {
 		epochID := e.writer.epochID
 		err := e.writer.rotate()
 		if err != nil {
 			return err
 		}
+		e.stateMu.Lock()
 		states := e.states
 		e.states = btree.New(8)
 		states.Ascend(func(i btree.Item) bool {
 			e.states.ReplaceOrInsert(i)
 			return true
 		})
+		e.stateMu.Unlock()
 		e.worker.taskCh <- rotateTask{epochID: epochID, states: states}
 	}
 	for _, op := range wb.raftLogOps {
 		e.writer.appendRaftLog(op)
+	}
+	e.mapMu.RLock()
+	for _, op := range wb.raftLogOps {
 		getRegionRaftLogs(e.entriesMap, op.regionID).append(op)
 	}
+	e.mapMu.RUnlock()
```

Comment:
In the `getRegionRaftLogs` function, if the `RegionRaftLogs` does not exist in the `entriesMap`, it will be created and stored in the map, and it may be created repeatedly with read lock.